[KIT-65] 임시 로그인 구현

glenn4105 opened the pull request:
# Pull Request Check List
- Major Reviewer: @dldhk97 @junhodo 

## CheckList

- [ ] Unit test coverage
- [ ] Backward compatibility
- [ ] Tested on dev/staging environment
- [ ] Documentation
- [x] Self reviewed

## Context

- datasource를 이용한 임시 로그인 구현
- dto 초기 구조 작성
- adapter layer (현. presentation layer) 수정

glenn4105 committed:
Feat: SignIn

- Dto 재사용 가능하도록 개선
- SignIn 요청 시 token 발급

# KIT-65

glenn4105 committed:
Commit Test

glenn4105 committed:
Feat: Authority

- 권한 검증을 위한 Authority feature 생성
- File server의 특성에 맞추어, 권한 그룹을 대상으로 검사하도록 설정

# KIT-65

glenn4105 committed:
Style: lombok, dto convention

- lombok, dto 컨벤션에 맞추어 수정
- 임시 작성되었던 파일 이동 및 삭제

# KIT-65

glenn4105 committed:
Feat: Security

- 인증 설정

# KIT-65

glenn4105 committed:
Feat: Account

- 기존 Api server의 account domain 이전

# KIT-65

glenn4105 committed:
Feat: Response 분리 / 이동

- Paging을 위하여 response 분리

# KIT-65

glenn4105 committed:
Fix: request 패키지 이동 및 명칭 변경

- CommonRequest.java -> BaseRequest.java
- app.dto -> app.dto.req

# KIT-65

glenn4105 committed:
Feat: 페이징 정렬 기준 추가

- 정렬 기준 속성 props 추가

# KIT-65

glenn4105 committed:
Feat: Presentation Layer

- Adapter Layer -> Presentation Layer
- presentation layer 이하 presenter 추가

# KIT-65

dldhk97 commented on src/main/java/com/se/fileserver/v1/account/application/service/AccountSignInService.java:
```diff
@@ -0,0 +1,38 @@
+  @Transactional
+  public String signIn(AccountSignInDto accountSignInDto, String ip){
+    Account account = accountRepository.findByIdString(accountSignInDto.getId()).orElseThrow(/*error*/);
+    if(account.isMatch(accountSignInDto.getPassword())){
```
조건문 앞에 느낌표가 들어가야 할 거 같습니다.

glenn4105 replied:
수정했습니다 감사합니다!

glenn4105 committed:
Setting: Lombok

glenn4105 committed:
Feat: Default Auth

- 누락된 Default Auth 추가

# KIT-65

junhodo commented on src/main/java/com/se/fileserver/v1/FileserverApplication.java:
```diff
@@ -2,12 +2,18 @@
+	@Bean
+	public PasswordEncoder passwordEncoder() {
+		return PasswordEncoderFactories.createDelegatingPasswordEncoder();
+	}
```
요거 암호화 알고리즘 뭐쓰는건가요?

glenn4105 replied:
초안이라 생각하고 작성하여 default인 bcrypt입니다.

junhodo commented on src/main/java/com/se/fileserver/v1/account/application/dto/AccountInfoDto.java:
```diff
@@ -0,0 +1,58 @@
+
+@Getter
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+public class AccountInfoDto {
```
Info는 불용어라 사용하지 않는게 좋을 것 같습니다, 필드를 보고 AccountDto와 AccountInfoDto 이름으로 비교하면 차이점이 없는 것 같네요

glenn4105 replied:
dto 컨벤션을 구성하면서 예시로 작성한 dto입니다. 혼동될 수 있으니 삭제하도록 하겠습니다.

junhodo commented on src/main/java/com/se/fileserver/v1/account/application/dto/request/AccountPaginationRequest.java:
```diff
@@ -0,0 +1,16 @@
+  public AccountPaginationRequest() {}
+
+  public AccountPaginationRequest(T dto, int page, int size,
+      Direction direction, String props) {
```
'props'라는 변수는 무엇인가요? direction을 설정할 필드 네임일 것 같은데 props는 추상적인것 같네요

glenn4105 replied:
원래 의미는 PageRequest 클래스의 properties입니다. 해당 속성명은 수정하겠습니다!

junhodo commented on src/main/java/com/se/fileserver/v1/account/application/service/AccountContextService.java:
```diff
@@ -0,0 +1,81 @@
+  private final AccountRepositoryProtocol accountRepositoryProtocol;
+  private final AuthorityGroupRepositoryProtocol authorityGroupRepositoryProtocol;
+
+  @Autowired
```
SpringBoot는 생성자 주입 생략가능한걸로 아는데 필요한 코드일까요? 아니면 생략하는 것도 좋을것 같네요

glenn4105 replied:
관련 어노테이션 삭제하겠습니다

junhodo commented on src/main/java/com/se/fileserver/v1/account/application/service/AccountContextService.java:
```diff
@@ -0,0 +1,81 @@
+  @Override
+  public UserDetails loadUserByUsername(String idString) throws UsernameNotFoundException {
+    Account account = accountRepositoryProtocol.findByIdString(idString)
+        .orElseThrow(() -> new RuntimeException(""));
```
엑셉션이 왜 없나요?

glenn4105 replied:
해당 내용 반영하겠습니다!

junhodo commented on src/main/java/com/se/fileserver/v1/account/domain/model/Account.java:
```diff
@@ -0,0 +1,80 @@
+@Getter
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@Entity
+public class Account extends BaseEntity {
```
추후에 password는 8~15자 특수문자 포함 등과 같은 도메인 로직이 포함되어야 합니다

glenn4105 replied:
entity라면 비밀번호가 인코딩 되어 들어가는 것으로 알고있는데 그런 경우에도 필요한가요?

junhodo replied:
@dubu4050 그럼 비밀번호는 dto 수준에서 검증하고 id, 닉네임 등에 대한 필드 검증이 필요합니다

glenn4105 committed:
Fix: 예외 처리

- 개선된 예외처리 컨벤션에 따라 각 로직에 예외처리 추가

# KIT-65

glenn4105 committed:
Fix: PaginationRequest 속성 명 변경

- props -> orderBy

# KIT-65

glenn4105 committed:
Fix: 임시 dto 삭제

- 컨벤션을 위한 임시 dto 삭제

# KIT-65

dldhk97 committed:
Merge branch 'main' into KIT-65]login

junhodo commented on src/main/java/com/se/fileserver/v1/account/domain/model/AccountVerifyToken.java:
```diff
@@ -0,0 +1,51 @@
+@Getter
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@Entity
+public class AccountVerifyToken extends BaseEntity {
```
파일서버에 존재할 필요가 있는 엔티티인가요?

glenn4105 replied:
추후에 파일 서버에서도 회원가입 기능이 추가될 수 있다 생각하여 api server로 부터 가져온 것입니다. 현재의 요구사항으로는 관계가 없는 것이니 빼는 것이 나을까요?

junhodo commented on src/main/java/com/se/fileserver/v1/account/infra/AccountDatasource.java:
```diff
@@ -0,0 +1,20 @@
+import org.springframework.stereotype.Component;
+
+@Component
+public class AccountDatasource {
```
entitiy 모델의 not null field는 모두 있어야하지 않을까요?

junhodo replied:
AccountDataSource는 언제 mysql로 바뀌어도 Account는 동일하게 동작해야합니다. 
필드 검증(not null field 등), 동일성과 동등성 보장, 영속화 등 
하지만 지금 당장 필요한 부분은 아니고 datasource를 여러 개 둘때는 각 datasource와 상관없이 도메인 로직이 동일하게 동작해야하는 것을 전달하고자 리뷰남깁니다.
지금은 not null field만 채워도 될것같네요

glenn4105 replied:
감사합니다 필요 속성 추가하겠습니다 😄

junhodo commented on src/main/java/com/se/fileserver/v1/account/presentation/presenter/AccountPresenterFormatter.java:
```diff
@@ -0,0 +1,14 @@
+import org.springframework.stereotype.Component;
+
+@Component
+public class AccountPresenterFormatter implements AccountPresenter{
```
Presenter/formatter 네이밍은 혹시 어떤걸 참고하셨나요?(그냥 궁금해서)

glenn4105 replied:
아래 링크를 참고했습니다
https://www.baeldung.com/spring-boot-clean-architecture
https://github.com/rcapraro/cleanarch

junhodo commented on src/main/java/com/se/fileserver/v1/authority/domain/model/AuthorityGroup.java:
```diff
@@ -0,0 +1,67 @@
+@Getter
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@Entity
+public class AuthorityGroup extends AccountGenerateEntity implements GrantedAuthority {
```
파일서버에 필요한가요?

glenn4105 replied:
해당 내용 확인 후 조치하겠습니다

junhodo commented on src/main/java/com/se/fileserver/v1/common/domain/exception/AttachmentTooLargeException.java:
```diff
@@ -0,0 +1,21 @@
+import org.springframework.http.HttpStatus;
+
+/**
+ * 사용자가 제한 크기를 초과한 파일을 첨부하려 하는 경우의 예외.
```
주석은 필요없을 것 같네요 네이밍이 좋아서 필요없어보입니다

dldhk97 replied:
넷플릭스 지니에서는 모든 예외 클래스에 주석을 달고가서, 한 줄 정도 주석이 들어가는 정도는 괜찮을 것 같습니다.

junhodo commented on src/main/java/com/se/fileserver/v1/common/domain/exception/BadRequestException.java:
```diff
@@ -0,0 +1,22 @@
+
+/**
+ * 잘못된 요청을 하였을 때 발생하는 에외.
+ * 서버가 요청을 이해할 수 없음을 의미합니다.
```
여기도 동일하게 필요없을 것 같습니다, 단 runtime exception과 exception이 분리가 되는 에러핸들링 구조라면 그것에 대한 컨텍스트가 공유되는 주석이 필요합니다

dldhk97 replied:
신규 에러 컨벤션에서는 Runtime Exception과 Checked Exception을 분리하지 않고 사용합니다.
[관련 문서](https://kit-se.atlassian.net/wiki/spaces/KITSE/pages/147030164?focusedCommentId=148209672#comment-148209672)

junhodo commented on src/main/java/com/se/fileserver/v1/common/domain/exception/InvalidInputException.java:
```diff
@@ -0,0 +1,21 @@
+import org.springframework.http.HttpStatus;
+
+/**
+ * 클라이언트가 요청한 리소스가 존재하지 않을 때 발생하는 예외.
```
동일하게 주석이 필요없어보입니다

junhodo commented on src/main/java/com/se/fileserver/v1/common/domain/exception/NotFoundException.java:
```diff
@@ -0,0 +1,23 @@
+import org.springframework.http.HttpStatus;
+
+/**
+ * 클라이언트가 요청한 리소스가 존재하지 않을 때 발생하는 예외.
```
`NotFoundException`보다 `ResourceNotFoundException`이 좋아보이네요, 
작성하신 주석에는 '리소스가 존재하지 않을 때'이지만 `NotFoundException`은 그 내용을 포함하지 못하는 것 같네요

glenn4105 replied:
확인 후 반영하겠습니다~

junhodo replied:
@dldhk97 se-api에도 동일하게 있나요?

dldhk97 replied:
@junhodo 기존 se-api에서는 모든 도메인에 대해 각각 따로 NO_SUCH_OOO 라는 네이밍으로 에러코드를 따로 만들어서 쓰고 있습니다.

junhodo commented on src/main/java/com/se/fileserver/v1/common/domain/exception/PermissionDeniedException.java:
```diff
@@ -0,0 +1,22 @@
+import org.springframework.http.HttpStatus;
+
+/**
+ * 인증된 사용자가 권한이 없는 리소스에 접근할 때 발생하는 예외.
```
여기도 주석이 필요없어보이는데, 권한쪽 로직이 파일서버에 필요한게 맞을까요?

glenn4105 replied:
권한 관련 내용은 확인 후 반영하겠습니다

junhodo commented on src/main/java/com/se/fileserver/v1/common/domain/exception/PreconditionFailedException.java:
```diff
@@ -0,0 +1,22 @@
+/**
+ * 메소드 호출시 전제조건이 충족되지 않은 경우의 예외.
+ */
+public class PreconditionFailedException extends SeException {
```
전제조건이 충족되지 않았다는 어떤걸 의미할까요? 개념이 넓어보이는데 사례를 들어주시면 좋을것같네용

glenn4105 replied:
`MethodArgumentNotValidException`, `ConstraintViolationException`이 발생했을 경우 사전 조건이 충족되지 않았다고 볼 수 있습니다

junhodo commented on src/main/java/com/se/fileserver/v1/common/domain/exception/UnauthenticatedException.java:
```diff
@@ -0,0 +1,23 @@
+/**
+ * 인증되지 않은 사용자가 인증이 필요한 리소스에 접근할 때 발생하는 예외.
+ */
+public class UnauthenticatedException extends SeException {
```
퍼미션 디나이와의 차이점이 무엇일까요?

glenn4105 replied:
permission denied: 사용자의 인증이 완료된 상황에서, 요청한 작업에 대해 수행할 권한이 없는 경우.
unauthorized: 인증이 누락되었거나, 과정에서 문제가 생긴 경우로 알고있습니다

junhodo commented on src/main/java/com/se/fileserver/v1/common/domain/exception/UniqueValueAlreadyExistsException.java:
```diff
@@ -0,0 +1,22 @@
+
+/**
+ * Unique value가 이미 존재하는 경우 발생하는 예외.
+ * 메시지에 이유를 포함하여 반환하기 바람니다.
```
여기도 네이밍이 좋아서 주석은 필요없어 보이네용

junhodo commented on src/main/java/com/se/fileserver/v1/common/domain/model/AccountGenerateEntity.java:
```diff
@@ -0,0 +1,20 @@
+@Component
+@MappedSuperclass
+@EntityListeners(AuditingEntityListener.class)
+public class AccountGenerateEntity extends BaseEntity{
```
`AccountGenerateEntity`는 이전에 제가 네이밍한 것 같은데 혹시 `AccountGeneratedEntity` 아니였나요?

glenn4105 replied:
api server main 브랜치 기준으로 `AccountGeneratedEntity`입니다
![image](https://user-images.githubusercontent.com/51086651/127272677-eea007d7-a26a-41be-beb2-d5ada2bfec85.png)

junhodo commented on src/main/java/com/se/fileserver/v1/account/domain/model/Account.java:
```diff
@@ -0,0 +1,80 @@
+    this.password = password;
+  }
+
+  public boolean isMatch(String plainPassword){
```
bycript를 사용하는 경우에는 이 인터페이스 구현하기 힘들 것 같네요! bycrypt는 암호화할 때 마다 결과가 다르게 나와서 bycrypt 내장함수를 사용해야 합니다

junhodo replied:
service layer에서 처리하거나 혹은 도메인에 인터페이스를 만든뒤에 한 번 래핑한 클래스를(adapter class) application layer에 구현해서 isMatch에서 passwordEncoder(domain에 정의된 인터페이스 구현체), inputPassword를 주입받는 것이 좋아보입니다
더 나아가면 Domain 서비스로 구현할 수도 있을것 같네요

glenn4105 replied:
감사합니다! 도메인 서비스 분리 및 passwordEncoder 내장함수 이용하여 수정하겠습니다

glenn4105 committed:
Fix: Dto Validation

- dto의 각 필드 validation 추가
- nested Object validation 설정

# KIT-65

glenn4105 committed:
Merge remote-tracking branch 'origin/KIT-65]login' into KIT-65]login

glenn4105 committed:
Fix: NotFoundException 명칭 수정

- ResourceNotFoundException로 명칭 수정

# KIT-65

glenn4105 committed:
Fix: verify entity 삭제

- Verify관련 entity는 필요하지 않을 것으로 보여 삭제

# KIT-65

glenn4105 committed:
Fix: password encoding

- passwordEncoder 내장 함수 이용하여 검증
- domain service 생성하여 구현

# KIT-65

glenn4105 committed:
Fix: Controller Advice 설정

- 예외 처리 컨벤션에 따라 설정

# KIT-65

glenn4105 committed:
Fix: Request 속성 및 함수 변경

- props -> orderBy
- PageRequest를 반환하도록 of 추가

# KIT-65

glenn4105 committed:
Fix: AccountDatasource 누락 속성 추가

- entity의 속성 중 누락되었던 속성 추가

# KIT-65

junhodo commented on src/main/java/com/se/fileserver/v1/account/domain/model/Account.java:
```diff
@@ -0,0 +1,83 @@
+@Getter
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@Entity
+public class Account extends BaseEntity implements AccountService {
```
왜 AccountService를 구현하나요?

glenn4105 replied:
도메인의 비즈니스 로직을 account service 인터페이스로 두어 isMatch와 같은 로직을 account 내부에서 구현해야 한다고 생각했습니다

junhodo replied:
음 제가 이전에 남긴 코멘트에 영향이 있는 것 같은데 지금 isMatch 함수는 
```
  @Override
  public boolean isMatch(PasswordEncoder passwordEncoder, String rawPassword) {
    return passwordEncoder.matches(rawPassword, this.password);
  }
```
와 같이 org.springframework.security.crypto.password.PasswordEncoder를 직접 이용합니다.
제가 남긴 코멘트의 의미는 Domain layer에서 framework 종속성을 없애야되기 때문에 DomainLayer에 PaswordEncoder 인터페이스를 재정의하고, Application Layer에org.springframework.security.crypto.password.PasswordEncoder를 주입받아 Wrapping하는 adapter Class가 필요하다는 의미였습니다.

지금 구조는 Domain Service의 AccountService를 Entity가 구현하고 있는데 그럼 Account는 Domain Service의 구현체이자 Entity의 역할을 수행하여 SRP 위반입니다(의미적으로도 어색합니다)

PasswordEncoder를 도메인 레이어로 끌고와 처리하는 것(도메인 로직으로서 처리하는 것)이 필수적이진 않습니다(다분히 Application layer에서 처리되어도 됩니다)

wonkajin commented on src/main/java/com/se/fileserver/v1/account/domain/model/Account.java:
```diff
@@ -0,0 +1,83 @@
+@Getter
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@Entity
+public class Account extends BaseEntity implements AccountService {
```
리뷰어가 아니라서, 코멘트 하나만 남깁니다.
이렇게 사용하면 4 레이어의 의미가 있나요?
엔티티가 서비스를 상속받는 구조는, 레이어를 해친다고 생각하네요.

glenn4105 replied:
account service는 비즈니스 로직을 정의한 인터페이스입니다. 도메인 레이어에 service 패키지 밑에 존재합니다. 때문에 레이어의 의미를 해친다고 생각하지는 않았습니다

junhodo replied:
Account Service가 Domain Service로서 구현하려했다는 의도를 이해했습니다~ 위 코멘트를 참고해주세용

glenn4105 committed:
Fix: domain service 삭제

- domain service(AccountService.java)삭제
- AccountSignInService 내 비밀번호 검증 로직 추가

# KIT-65

glenn4105 committed:
Fix: Default Authority User 파라미터 변경

- yml의 ANONYMOUS ID/PW를 사용하도록 변경

# KIT-65

dldhk97 committed:
Merge branch 'main' into KIT-65]login
